pydantic/main.py

```diff
@@ -57,6 +64,97 @@
+        return origin is typing.Union or origin is types.UnionType
+
+else:
+
+    def _is_union(origin):
+        return origin is typing.Union
+
+
+def _handle_union(args, value):
+    # TODO: union_mode/discriminators?
+    for possible_type in args:
+        result = _recursive_model_construct(possible_type, value)
+        if result != value:
+            return result
+
+    # Otherwise, no union member matched
+    return value
+
+
+def _recursive_model_construct(annotation: type | None, value: Any):
+    # No annotation? Early exit
+    if annotation is None:
+        return value
+
+    # If value is already a BaseModel, just return that
+    # NOTE: if we see this, we assume that all of the children of the supplied BaseModel have been recursed through and
+    # converted to BaseModels, OR that the user manually provided this value and would like us to *not* recurse through
+    # its children and instead leave it "as-is". Depending on the circumstance, the user might want either behavior, so
+    # perhaps this could be toggled with another parameter/setting.
+    if isinstance(value, BaseModel):
+        return value
+
+    # Try to get a "parent" class if we're dealing with a structural type like list, dict, sequence, etc.
+    # Otherwise just treat the annotation itself as the parent
+    origin = get_origin(annotation)
+    origin = annotation if origin is None else origin
+
+    # Get all of the classes in the brackets, if there are any
+    args = get_args(annotation)
+
+    # Union and Literal don't play well with isinstance/issubclass, so we check them early
+    if origin is Literal:
+        return value
+    elif _is_union(origin):
+        return _handle_union(args, value)
+    # Then we check to see if the annotation describes it as a Base/RootModel
+    elif issubclass(origin, BaseModel):
+        if origin.__pydantic_root_model__:  # my saviour
+            return origin.model_construct(value, _recursive=True)  # treat like RootModel
+        else:
+            try:
+                return origin.model_construct(**value, _recursive=True)  # treat like BaseModel
+            except TypeError:
+                return value
+    # Then we check for subclasses of mappings/sequences/sets
+    # But if there are no arguments to the annotation, we have to treat the value generically anyway
+    elif len(args) == 0:
+        return value
+    elif issubclass(origin, typing.Tuple):
+        if isinstance(args[-1], type(Ellipsis)):  # tuple[T, ...]
+            member_type = args[0]
+            return type(value)(_recursive_model_construct(member_type, v) for v in value)
+        else:  # tuple[A, B, C]
+            return type(value)(_recursive_model_construct(member_type, value[i]) for i, member_type in enumerate(args))
+    elif issubclass(origin, typing.Sequence) and not issubclass(origin, (str, bytes)):
+        member_type = args[0]
+        return [_recursive_model_construct(member_type, v) for v in value]
```

Comment:
If that's the case, then presumably the "correct" way to do it would be to modify `value` and return that... but that only works if the Sequence/Mapping is mutable, by definition. In immutable cases in order to be correct we're going to have to create a new instance of the object; if we can't do something like `type(value)(...)` then I have no idea how to dynamically do it. How would it be done on the rust side of things?